[allocator] Generalize recording to a pool

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96542

Previously the allocator would query whether a stream was recording a graph,
and look up the pool associated with a graph. This change has the allocator
directly associate a stream with a mempool, decoupling "record this stream to a pool"
from the action of "record all actions to a cuda graph".

zdevito committed:
[allocator] Generalize recording to a pool

Previously the allocator would query whether a stream was recording a graph,
and look up the pool associated with a graph. This change has the allocator
directly associate a stream with a mempool, decoupling "record this stream to a pool"
from the action of "record all actions to a cuda graph".

[ghstack-poisoned]

eellison commented on aten/src/ATen/cuda/CUDAGraph.cpp:
```diff
@@ -290,7 +289,7 @@ void CUDAGraph::reset() {
   if (has_graph_ || has_graph_exec_) {
     // notifyCaptureDestroy may throw. How should we handle this?
-    c10::cuda::CUDACachingAllocator::notifyCaptureDestroy(capture_dev_, mempool_id_);
+    c10::cuda::CUDACachingAllocator::destroyPool(capture_dev_, mempool_id_);
```
IMO, `destroyPool` is a little misleading. We're notifying that the current capture is destroyed, but there might be other references to the pool that keep it alive.

eellison commented on c10/cuda/CUDAMallocAsyncAllocator.cpp:
```diff
@@ -762,30 +774,15 @@ struct CudaMallocAsyncAllocator : public CUDAAllocator {
-      free_impl(it);
-    }
-
-    ungraphed_ptrs_defer_free_until_no_capture.clear();
```
is this logic not important ?

zdevito replied:
I moved it to `mallocAsync`: in other words once we are done graph capturing but before we do further allocations we free these buffers. It will end up being very similar to what happened before but allows us to have a more meaningful API for what is happening with pools (allocator knows about the intent to create and allocate into pools but doesn't need to know the specifics of the cuda graph API).

eellison reviewed (approved):
Sweet ! Looks good to me, but IMO we could rename the api a bit, since it won't necessarily destroy the pool. Also had one question about the removed `Avoid freeing uncaptured ptrs during CUDA graph capture` logic

You are failing lint fyi

zdevito committed:
Update on "[allocator] Generalize recording to a pool"


Previously the allocator would query whether a stream was recording a graph,
and look up the pool associated with a graph. This change has the allocator
directly associate a stream with a mempool, decoupling "record this stream to a pool"
from the action of "record all actions to a cuda graph".

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge
